File: src/offerings/booking.py
import uuid
from offerings.offering import Offering
from users.client import Client
from datetime import date, datetime 

class Booking:
    def __init__(self, client: Client, offering: Offering):
        self.offering = offering
        self.client = client
        self.booking_id = str(uuid.uuid4())
        self.date = date.today()
        self.status = True
        self.time = datetime.now()  
        self.guardian = None
        self.timeslot = None

    def make_Booking(self, client, offering, time_slot):
        if offering.check_available():
            self.client = client
            self.offering = offering
            self.time_slot = time_slot
            self.status = True  
            offering.update_availability(False)  
            return True
        else:
            print("Offering is not available.")
            return False
        

    def cancel_booking(self):
        self.status = False
        self.offering.update_availability(True)
        return True
    
    def isAvailable(self):
        return self.status
    
    def get_booking_details(self):

        details = (
        f"Booking ID: {self.booking_id}\n"
        f"Client Name: {self.client.name}\n"
        f"Client ID: {self.client.id}\n"
        f"Offering Location: {self.offering.get_location()}\n"
        f"Lesson Type: {self.offering.lesson_type}\n"
        f"Date: {self.date}\n"
        f"Time Slot: {self.time_slot}\n"
        f"Status: {'Active' if self.status else 'Canceled'}"
    )
        if self.guardian:
            details += f"\nGuardian: {self.guardian.name} (ID: {self.guardian.id})"
        return details

    def update_status(self, new_status: bool):
        self.status = new_status

    def set_guardian(self, guardian: Client):
        self.guardian = guardian

    def change_time_slot(self, new_time_slot):
        self.time_slot = new_time_slot

File: src/main/java/offerings/Booking.java
package offerings;

import java.util.Date;
import java.util.UUID;
import users.Client;

public class Booking {

    private String bookingId;
    private Client client;
    private Offering offering;
    private Date date;
    private boolean status;
    private Date time;
    private Client guardian;
    private String timeSlot;

    // Constructor
    public Booking(Client client, Offering offering) {
        this.bookingId = UUID.randomUUID().toString();
        this.client = client;
        this.offering = offering;
        this.date = new Date();
        this.status = true;
        this.time = new Date();  // Current date and time
        this.guardian = null;
        this.timeSlot = null;
    }

    // Method to make a booking
    public boolean makeBooking(Client client, Offering offering, String timeSlot) {
        if (offering.checkAvailable()) {
            this.client = client;
            this.offering = offering;
            this.timeSlot = timeSlot;
            this.status = true;
            offering.updateAvailability(false);
            return true;
        } else {
            System.out.println("Offering is not available.");
            return false;
        }
    }

    // Method to cancel a booking
    public boolean cancelBooking() {
        if (this.status) {
            this.status = false;
            this.offering.updateAvailability(true);
            return true;
        } else {
            System.out.println("Booking is already canceled.");
            return false;
        }
    }

    // Method to check if the booking is active
    public boolean isAvailable() {
        return this.status;
    }

    // Method to get booking details
    public String getBookingDetails() {
        StringBuilder details = new StringBuilder();
        details.append("Booking ID: ").append(bookingId).append("\n")
                .append("Client Name: ").append(client.getName()).append("\n")
                .append("Client ID: ").append(client.getId()).append("\n")
                .append("Offering Location: ").append(offering.getLocation()).append("\n")
                .append("Lesson Type: ").append(offering.getLessonType()).append("\n")
                .append("Date: ").append(date).append("\n")
                .append("Time Slot: ").append(timeSlot).append("\n")
                .append("Status: ").append(status ? "Active" : "Canceled");

        if (guardian != null) {
            details.append("\nGuardian: ").append(guardian.getName())
                    .append(" (ID: ").append(guardian.getId()).append(")");
        }

        return details.toString();
    }

    // Method to update the booking status
    public void updateStatus(boolean newStatus) {
        this.status = newStatus;
    }

    // Method to set a guardian for underage clients
    public void setGuardian(Client guardian) {
        this.guardian = guardian;
    }

    // Method to change the time slot of the booking
    public void changeTimeSlot(String newTimeSlot) {
        this.timeSlot = newTimeSlot;
    }
}
